src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -0,0 +1,178 @@
+    #: argument is described by an entry in the Dict.
+    _optional_args = {}
+    # The positional arguments to allocate must all be References (or
+    # ArrayReferences but they are a subclass of Reference).
+    _required_args[Intrinsic.ALLOCATE.name] = ArgDesc(1, None, Reference)
+    _optional_args[Intrinsic.ALLOCATE.name] = {"mold": Reference,
+                                               "stat": Reference}
+    _required_args[Intrinsic.DEALLOCATE.name] = ArgDesc(1, None, Reference)
+    _optional_args[Intrinsic.DEALLOCATE.name] = {"stat": Reference}
+    _required_args[Intrinsic.RANDOM.name] = ArgDesc(1, 1, Reference)
+    _optional_args[Intrinsic.RANDOM.name] = {}
+
+    @classmethod
+    def create(cls, intrinsic, arguments):
+        '''Create an instance of class cls given the type of intrinsic and a
+        list of nodes (or name-and-node tuples) for its arguments. Any
+        named arguments *must* come after any required arguments.
+
+        :param intrinsic: the Intrinsic being called.
+        :type intrinsic: py:class:`psyclone.psyir.IntrinsicCall.Intrinsic`
+        :param arguments: the arguments to this routine, and/or \
+            2-tuples containing an argument name and the \
+            argument. Arguments are added as child nodes.
+        :type arguments: List[ \
+            Union[:py:class:``psyclone.psyir.nodes.DataNode``, \
+                  Tuple[str, :py:class:``psyclone.psyir.nodes.DataNode``]]]
+
+        :returns: an instance of cls.
+        :rtype: :py:class:`psyclone.psyir.nodes.IntrinsicCall`
+
+        :raises TypeError: if any of the arguments are of the wrong type.
+        :raises ValueError: if any optional arguments have incorrect names \
+            or if a positional argument is listed *after* a named argument.
+        :raises ValueError: if the number of supplied arguments is not valid \
+            for the specified intrinsic.
+
+        '''
+        if not isinstance(intrinsic, cls.Intrinsic):
+            raise TypeError(
+                f"IntrinsicCall create() 'intrinsic' argument should be an "
+                f"instance of IntrinsicCall.Intrinsic but found "
+                f"'{type(intrinsic).__name__}'.")
+
+        if not isinstance(arguments, list):
+            raise TypeError(
+                f"IntrinsicCall.create() 'arguments' argument should be a "
+                f"list but found '{type(arguments).__name__}'")
+
+        if cls._optional_args[intrinsic.name]:
+            optional_arg_names = sorted(list(
+                cls._optional_args[intrinsic.name].keys()))
+        else:
+            optional_arg_names = []
+
+        # Validate the supplied arguments.
+        reqd_args = cls._required_args[intrinsic.name]
+        opt_args = cls._optional_args[intrinsic.name]
+        last_named_arg = None
+        pos_arg_count = 0
+        for arg in arguments:
+            if isinstance(arg, tuple):
+                name = arg[0].lower()
```

Comment:
Maybe a TypeError if `not isinstance(arg[0], str)` here?

Reply:
Yes, done.